Add caltech HPC environment file

moxcodes opened the pull request:
## Proposed changes

Adds a new environment file for compiling natively on the caltech HPC system with gcc, based heavily on the frontera environment file. This also adds an extra linker flag that fixed a linker error that appeared for that system.


### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [ ] Code
- [ ] Documentation
- [x] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

moxcodes committed:
Add caltech HPC environment file

kidder commented:
@fmahebert can you provide a review on this?

geoffrey4444 commented on support/Environments/caltech_hpc_gcc.sh:
```diff
@@ -0,0 +1,284 @@
+#!/bin/env sh
+
+# Distributed under the MIT License.
+# See LICENSE.txt for details.
```
Could you please add docs on how to build here or in the dev guide? E.g., you have to install the modules, and you have to compile on a compute node. I get out of memory errors when trying to build on the head node, and so far, I only figured out how to get a single core compute node (or messed something else up, so it only builds in serial).

geoffrey4444 commented on support/Environments/caltech_hpc_gcc.sh:
```diff
@@ -0,0 +1,284 @@
+
+# Distributed under the MIT License.
+# See LICENSE.txt for details.
+
```
I get a bunch of warnings from boost when I compile...can these be silenced?

geoffrey4444 commented on support/Environments/caltech_hpc_gcc.sh:
```diff
@@ -0,0 +1,284 @@
+#!/bin/env sh
```
Could you please also add a submit script to support/SubmitScripts?

geoffrey4444 reviewed (changes requested):
Thanks for doing this!!

I'm trying this out, but it's painfully slow...I can only compile on compute nodes, and for some reason, it's dividing up my tasks among a single core, even though I asked for 1 node with 32 tasks per node in my debug job. Docs on how to build most efficiently would be most welcome!

This might be beyond the scope of this PR, but I wonder if you could add a version that builds with llvm, since llvm builds so much faster than gcc does?

moxcodes commented:
@geoffrey4444 I agree, the compilation is agonizingly slow with the gcc from this environment file. I spent a little bit of time trying to get the clang version to work, but didn't manage to get it working on a first pass.  I can make another attempt, but it would likely be another environment ~variable~ file (analogous to wheeler).

geoffrey4444 commented:
> @geoffrey4444 I agree, the compilation is agonizingly slow with the gcc from this environment file. I spent a little bit of time trying to get the clang version to work, but didn't manage to get it working on a first pass.  I can make another attempt, but it would likely be another environment ~variable~ file (analogous to wheeler).

I think a separate PR is reasonable for clang.

Can you compile on a compute node with multiple cores? I tried, but checking too showed it was dividing the tasks onto 1 core, even though I asked slurm for many tasks per node

moxcodes commented:
@geoffrey4444 I was able to build in parallel on a compute node when I got an interactive node with the command:
```
srun -t 02:00:00 -N 1 -c 32 --mem=192000 /bin/bash
```
When I tried it this gives a 'blank' shell, which is a little hard to navigate, but once you've been allocated the job you can also just ssh in to the compute node. When I compile in that environment, I saw several cc1plus tasks appearing in top.

geoffrey4444 commented:
> @geoffrey4444 I was able to build in parallel on a compute node when I got an interactive node with the command:
> 
> ```
> 
> srun -t 02:00:00 -N 1 -c 32 --mem=192000 /bin/bash
> 
> ```
> 
> When I tried it this gives a 'blank' shell, which is a little hard to navigate, but once you've been allocated the job you can also just ssh in to the compute node. When I compile in that environment, I saw several cc1plus tasks appearing in top.

I did also see several processes running, but all ran such that compiled the usage added up to 100%, so it wasn't letting me use multiple cores

moxcodes commented:
@geoffrey4444 did you compile in the separate ssh shell or the weird undecorated shell that was provided in answer to the srun command? When I did the latter, the top snapshot I have from the simultaneous ssh connection shows at least two compilation processes hitting ~99% cpu.

geoffrey4444 commented:
> @geoffrey4444 did you compile in the separate ssh shell or the weird undecorated shell that was provided in answer to the srun command? When I did the latter, the top snapshot I have from the simultaneous ssh connection shows at least two compilation processes hitting ~99% cpu.

Neither: I just tried a modified version of your command

`srun -t 02:00:00 -N 1 -c 32 --mem=192000 -A sxs -D . --pty /bin/bash`

and that one gave me a regular shell that does let me use multiple cores. So now it's compiling. I guess my original attempt had some incorrect options to srun. Thanks for your help!

nilsdeppe commented:
Should there be some documentation added to the `Installation on Clusters` part of the dox? @geoffrey4444 thoughts? This sounds like it's basically good to go other than clarifications on how to deal with system-specific things, which should go into `Installation On Clusters`.

nilsdeppe commented:
@geoffrey4444 ping on my question

geoffrey4444 commented:
I'm in favor of docs on the cluster, since currently tests fail when I try to run them with ctest

moxcodes commented:
I'll investigate the error @geoffrey4444 found and update the docs for the cluster before we try to merge this PR.

moxcodes committed:
Remove extra quotes from spectre test cases

moxcodes committed:
Add CaltechHPC to installation on clusters dox

moxcodes commented:
Updated this with a few commits that fix the errors @geoffrey4444 pointed out. It should build and successfully test now.
I've tested it in the docker and on wheeler, to make sure that the cmake changes don't make things fail on other systems.
Cheers!

nilsdeppe commented:
@moxcodes this needs a rebase. I pinged @geoffrey4444 on Slack to test

kidder commented:
ping @geoffrey4444

moxcodes commented:
Rebased.
Cheers!

nilsdeppe commented:
@geoffrey4444 can you test this out please?

kidder commented:
ping @geoffrey4444

kidder commented:
ping @geoffrey4444   @moxcodes this needs a rebase

moxcodes commented:
rebased
Cheers!

moxcodes committed:
Add CaltechHpc.sh submit script

It is very similar to the Ocean submit script, because charm is built
in a similar way on Caltech HPC.

moxcodes commented:
While this is waiting to be reviewed, I added a CaltechHpc.sh submit script that is very similar to the ocean one, but because that similarity is not necessarily obvious to users, it seems nice to have.

moxcodes commented:
Updated the submit script with the environment fix from #2860.

Sizheng-Ma commented on support/SubmitScripts/CaltechHpc.sh:
```diff
@@ -0,0 +1,92 @@
+
+# These commands load the relevant modules and cd into the run directory,
+# creating it if it doesn't exist
+source ${SPECTRE_HOME}/support/Environments/caltech_hpc.sh
```
This should be `source ${SPECTRE_HOME}/support/Environments/caltech_hpc_gcc.sh` in order to be consistent.

Sizheng-Ma commented on support/SubmitScripts/CaltechHpc.sh:
```diff
@@ -0,0 +1,92 @@
+# spectre environment between the start of the remote session and starting the
+# spectre executable
+echo "#!/bin/sh
+source /home/moxon/spectre/support/Environments/ocean_clang.sh
```
This line should be `source /home/moxon/spectre/support/Environments/caltech_hpc_gcc.sh` for Caltech HPC.

moxcodes commented:
fixup posted for the caltech HPC submit script. Thanks for the review, @Sizheng-Ma!

Sizheng-Ma commented:
Those changes look good to me.

nilsdeppe commented:
@moxcodes do you want to rebase+squash this and then I can look over it so we can get this merged?

moxcodes commented:
@nilsdeppe Yeah, I'll do that soon. I want to include jemalloc install on this because it makes a big difference, but that'll be pretty quick once I've run my tests.

nilsdeppe commented:
okay, great. Thanks!

moxcodes commented:
Squashed in the jemalloc change and a version bump to blaze.
Cheers!

nilsdeppe commented on tests/Unit/NumericalAlgorithms/Spectral/Test_Spectral.cpp:
```diff
@@ -131,7 +131,6 @@ void test_weak_differentiation() {
               approx(weak_diff_matrix(i, l)));
       }
     }
-
```
the commit message seems reasonable, but it doesn't seem to reasonably describe the commit :)

moxcodes replied:
Hah. oops. I guess that change got merged separately, aside from the newline delete. Dropped the commit.

nilsdeppe reviewed (changes requested):
LGTM, there is one unusual commit, please squash whatever you need to do with it :)

nilsvu commented on support/Environments/caltech_hpc_gcc.sh:
```diff
@@ -0,0 +1,312 @@
+        tar xzf v6.10.2.tar.gz
+        mv charm-6.10.2 charm
+        cd $dep_dir/charm
+        ./build charm++ verbs-linux-x86_64-smp --with-production -j4
```
I think you can delete this line (it's redundant with the next line)

moxcodes replied:
I'm unsure about what is intended from charm regarding that, actually -- the help text in the build script made me concerned -- from LIBS was `compile additional parallel libraries with Charm++ core`, and the 'with' there seemed ambiguous about whether it uses the charm core to make the compile the additional libs, or if it intended 'in addition'. I guess it seems to work with just the `LIBS` build, so maybe that's fine?
I included this mostly for safety in case that interpretation wasn't entirely correct.

nilsvu commented on support/Environments/caltech_hpc_gcc.sh:
```diff
@@ -0,0 +1,312 @@
+          -D CMAKE_Fortran_COMPILER=gfortran \
+          -D MEMORY_ALLOCATOR=JEMALLOC \
+          -D BUILD_PYTHON_BINDINGS=off \
+          -D PYTHON_EXECUTABLE=`which python3` \
```
This should be `Python_EXECUTABLE`. Not sure if cmake has some failsafes in place to make this capitalization work as well.

nilsvu commented on tests/Unit/CMakeLists.txt:
```diff
@@ -64,6 +64,7 @@ target_link_libraries(
   Options
   SystemUtilities
   Utilities
+  util
```
I can't find this target anywhere. What is it?

moxcodes replied:
I put it in because when I first tried this on Caltech HPC the linker failed because `libutil` was needed for something in Pypp, I think.
Now it no longer fails without this, likely because other linker problems were fixed and this links in via other dependencies -- I'll remove this and fix the `Python_EXECUTABLE` thing in a follow-up.
